### DISCUSSION TITLE
Build of dll library on windows

### DISCUSSION THREAD

**Discussion by Josko692** [JSBSim-Team/jsbsim#718] [Q&A] [answered]
Hello, I work on windows 10 with mingw64 tools (gcc, g++, mingw32-make). 
When I passed to the v1.1.12 of JSBSim I decided to use the shared Library. 
I use the CMakeLists.txt of the project. My cmake line is : 

cmake .. -G"MinGW Makefiles" -DCMAKE_CXX_FLAGS_DEBUG="-g -Wall" -DCMAKE_C_FLAGS_DEBUG="-g -Wall" -DCMAKE_BUILD_TYPE=Debug -DBUILD_SHARED_LIBS=ON -DBUILD_PYTHON_MODULE=OFF -DCMAKE_INSTALL_PREFIX=../install

There is no problem with the compilation of the dll but when linking. Apparently there is a problem with the FGOutput files (in src/models/) and I can't link to get JSBSim.exe. The script stops on an error and does not go to the end. 
Did anyone manage to compile/link the V1.1.12 version using the dll ? 

I attach a capture with the errors. Thank you for your help.
![Capture](https://user-images.githubusercontent.com/101246645/189353934-95000d86-853a-4975-b29c-1729fdc1083f.png)

**Comment by bcoconni** [comment]
@Josko692 I can reproduce the bug so I opened the issue #719.

**Reply by bcoconni** [reply]
I'll submit a PR shortly to fix this issue. Meanwhile, you should be able to succeed compiling the DLL with MinGW32 by adding `JSBSIM_API` to the class header in `src/models/FGOutput.h`:

```diff
diff --git a/src/models/FGOutput.h b/src/models/FGOutput.h
--- a/src/models/FGOutput.h
+++ b/src/models/FGOutput.h
@@ -122,7 +122,7 @@ CLASS DOCUMENTATION
 CLASS DECLARATION
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%*/

-class FGOutput : public FGModel
+class JSBSIM_API FGOutput : public FGModel
 {
 public:
   FGOutput(FGFDMExec*);
```

**Reply by Josko692** [reply]
Thank you for your answer and your reactivity. I hope that next time I will be able to find the solution and propose a pull request.

**Comment by seanmcleod** [comment]
Hmm, @bcoconni, @imkogane, @jeffneosim, @gallonmate I wonder if this lack of `JSBSIM_API` for `FGOutput`, although not causing a linker issue with the Microsoft linker is possibly causing a runtime issue in terms of this report with regards to not being able to generate an output log file - https://github.com/JSBSim-Team/jsbsim/issues/711#issuecomment-1235284201

**Reply by bcoconni** [reply]
Indeed, what puzzles me is that the DLL linkage succeeds for `JSBSim.exe` with MSVC even though `JSBSIM_API` is missing 😲. Actually, MSVC issues **a whole lot** of [warning C4251 messages](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4251?view=msvc-170) when compiling JSBSim as a DLL but nothing specifically related to `FGOutput`.

Microsoft docs mention that the purpose of warnings C2451 is "*to minimize the possibility of data corruption when exporting a class declared as `__declspec(dllexport)`*". So the next step might possibly be fixing the warnings C2451 ?!?

**Reply by gallonmate** [reply]
Nice thought/connection. I never tested UE5 logging until now. Though I can confirm output logging works without any changes related to this dll bug. For Unreal Engine logging, the aircraft xml file just needed to specify some folder location to work:
` <output name="F:\Downloads\c172p\JSBout172C.csv" type="CSV" rate="60"> `

**Comment by seanmcleod** [comment]
@jeffneosim I moved your comment to the existing discussion https://github.com/JSBSim-Team/jsbsim/issues/711#issuecomment-1242712041 since there was nothing in the comment that was specific to this `FGOutput` issue plus there is an existing discussion that covers your comment.

**Comment by ghost** [comment]
Thanks, This discussion is the one from my students, and I just write here it was faster for me….

**Comment by seanmcleod** [comment]
Sure it's faster for you, but it doesn't help anyone else looking into this specific issue with `FGOutput` now, or any other JSBSim users who in the future do a search for `FGOutput` issues etc.

**Comment by bcoconni** [comment, marked as answer]
@Josko692 The bug should now be fixed in the `master` branch (commit 9a63f2de8074f9f9ede15c561a271eb4648e3b68).
